pydantic_graph/pydantic_graph/nodes.py

```diff
@@ -0,0 +1,162 @@
+from __future__ import annotations as _annotations
+
+from abc import ABC, abstractmethod
+from dataclasses import dataclass, is_dataclass
+from functools import cache
+from typing import TYPE_CHECKING, Any, ClassVar, Generic, get_origin, get_type_hints
+
+from typing_extensions import Never, TypeVar
+
+from . import _utils, exceptions
+
+if TYPE_CHECKING:
+    from .state import StateT
+else:
+    StateT = TypeVar('StateT', default=None)
+
+__all__ = 'GraphContext', 'BaseNode', 'End', 'Edge', 'NodeDef', 'RunEndT', 'NodeRunEndT'
+
+RunEndT = TypeVar('RunEndT', default=None)
+"""Type variable for the return type of a graph [`run`][pydantic_graph.graph.Graph.run]."""
+NodeRunEndT = TypeVar('NodeRunEndT', covariant=True, default=Never)
+"""Type variable for the return type of a node [`run`][pydantic_graph.nodes.BaseNode.run]."""
+
+
+@dataclass
+class GraphContext(Generic[StateT]):
+    """Context for a graph."""
+
+    state: StateT
+    """The state of the graph."""
+
+
+class BaseNode(ABC, Generic[StateT, NodeRunEndT]):
```

Comment:
I wanted to find out if we really need this type NodeRunEndT or if we can just use only RunEndT.

It appears that within the Graph class all the methods are using RunEndT and NodeRunEndT is only used in nodes.py but in graph.py all the references for this type are RunEndT.

It is just my observation so I wanted to confirm if NodeRunEndT is really needed.

Thanks.

Reply:
they have different meanings, so I think it's fine to have different versions.